### PR TITLE
Fix uTransport send() definition to map source.

### PR DESCRIPTION
**Pull request by stevenhartley**
#65

### CODE REVIEW

**Review by sophokles73 — COMMENTED** [review]
This is now consistent with the protobuf file. However, it feels very strange that when invoking the `send()` method, I need to specify the source instead of the destination that I want to send the message to. Intuitively, the destination address seems to be the most important parameter while only some messages will require including an origin (source) address, e.g. for specifying the reply-to address. So, from a developer's perspective I would expect to find the destination address as the first parameter to the `send` method and the reply-to/source address as an (optional) property of the UAttributes. Is there a particular reason why you did it the other way around?

**Comment by stevenhartley** [comment]
@sophokles73 understood and agree, will revise based on our discussion.

File: up-l1/README.adoc [hunk 1]
```diff
@@ -87,32 +87,24 @@ For example, if HartleyTransport can map only UPriority to its PDU header, then
 
 The API signature:
 
-`OUT UStatus send(UUri&, IN UPayload&, IN UAttributes&)`
+`OUT UStatus send(UMessage&)`
```
Comment (PLeVasseur) [inline up-l1/README.adoc:90]: :+1:

File: up-l1/README.adoc [hunk 2]
```diff
@@ -67,7 +67,7 @@ NOTE: uProtocol data model used by the uTransport interface are declared in http
 `UListener` is an interface used for receiving messages by the client uE from the uTransport layer. Listeners are used for receiving messages asynchronously by the uTransport layer.  
 
   * Type name *MUST* be `UListener`
-  * *MUST* contain the method `UStatus onReceive(UUri&, UPayload&, UAttributes&)` that is called by the Transport to notify (callback) the client
+  * *MUST* contain the method `UStatus onReceive(UMessage&)` that is called by the Transport to notify (callback) the client
```
Comment (sophokles73) [inline up-l1/README.adoc:70]: does this mean that the listener can return a `UStatus` to the transport implementation? Is there a particular expectation how that should be used and what the transport implementation should do about it?

**Review by sophokles73 — COMMENTED** [review]
IMHO you should also adapt the signature of `Receive()` accordingly.

File: up-l1/README.adoc (hunk 2 above)
Comment (stevenhartley) [reply up-l1/README.adoc:70]: @sophokles73 this has been a long standing debate and there are two school of thoughts from developers, some want to return (so the operations can be part of a flow) and others say "it is a callback, it shouldn't do anything". I would love to hear others opinions but to this date there is no defined behavior as to what we should do if the return is anything but OK.

Comment (stevenhartley) [reply up-l1/README.adoc]: Done, I checked and up-java returns void anyways

**Comment by tamarafischer** [comment]
@sophokles73 and people, uProtocol mission statement is connecting uE everywhere and anywhere.
We are using service oriented architecture with strong contracts using the protobuf files and we are optimizing on message sending using pub/sub patterns.
In the pub/sub patterns the sender **does not care** about receivers - it just sends out messages.
It can optimize that is no one is subscribed, nothing is sent.

" I need to specify the source instead of the destination that I want to send the message to" - correct! This is to enable routers/dispatchers to implement the mission statement of connecting everyone everywhere.

An owner of data (service oriented) can publish a state change - the platforms job is to deliver this state change to interested parties. Interested parties register as such by subscribing to said state change topic.

Routers and dispatchers are interested in the source since their job is knowing who is subscribed, not the owners of the data.

Separation of concerns in this way, gives the protocol all the wiggle room it is needs to optimize each system delivery.

**Comment by tamarafischer** [comment]
If RPC is what is confusing, then I have a ton to say about that and it is more than a comment typing. But it should be solved in the RPC part and not in the actual UTransport part.

**Comment by sophokles73** [comment]
@tamarafischer there is no dissent about what you stated. I also agree with

> In the pub/sub patterns the sender does not care about receivers - it just sends out messages.

and I would be inclined to append

> to an address.

because in the end, the receiver(s) of the message are represented by this address, regardless of whether it is zero, one or multiple receivers.

I also believe that when it comes to defining a developer-facing API, we should try to make the API as easy to use and comprehensible as possible. While I can easily see that a message is being published/sent-out *to* an address/topic/receiver (whatever you might want to call it), I have a hard time understanding why I need to put the address that I want to send the message *to* into a parameter that is called *source* (even if at the transport level, calling it like that might make sense).

Moving the *source* parameter into `UAttributes` does not really solve this issue FMPOV, but I guess we can alleviate this problem by means of the `UAttributesBuilder` factory methods which we can use to hide the mapping of the publish address to the corresponding `UAttributes` property.

**Comment by stevenhartley** [comment]
> @tamarafischer there is no dissent about what you stated. I also agree with
> 
> > In the pub/sub patterns the sender does not care about receivers - it just sends out messages.
> 
> and I would be inclined to append
> 
> > to an address.
> 
> because in the end, the receiver(s) of the message are represented by this address, regardless of whether it is zero, one or multiple receivers.
> 
> I also believe that when it comes to defining a developer-facing API, we should try to make the API as easy to use and comprehensible as possible. While I can easily see that a message is being published/sent-out _to_ an address/topic/receiver (whatever you might want to call it), I have a hard time understanding why I need to put the address that I want to send the message _to_ into a parameter that is called _source_ (even if at the transport level, calling it like that might make sense).
> 
> Moving the _source_ parameter into `UAttributes` does not really solve this issue FMPOV, but I guess we can alleviate this problem by means of the `UAttributesBuilder` factory methods which we can use to hide the mapping of the publish address to the corresponding `UAttributes` property.

@tamarafischer we are not changing the definition of source and sink, just making the transport API more intuitive.

**Pull request merged**